lib/mongodb.nit

```diff
@@ -0,0 +1,708 @@
+	# assert json["ELS"].as(JsonArray).is_empty
+	# ~~~
+	fun to_json: JsonObject do return to_s.parse_json.as(JsonObject)
+
+	# Utility method to set `Sys.last_mongoc_error`.
+	private fun set_mongoc_error(err: BSONError) do sys.last_mongoc_error = err
+end
+
+# Introduces BSON services.
+redef class JsonObject
+
+	# Init `self` from a BSON object.
+	#
+	# ~~~
+	# var str = """{ "age" : 10, "name" : "Rick", "ELS" : [  ] }"""
+	# var bson = new BSON.from_json_string(str)
+	# var json = new JsonObject.from_bson(bson)
+	# assert json["age"] == 10
+	# assert json["name"] == "Rick"
+	# assert json["ELS"].as(JsonArray).is_empty
+	# ~~~
+	init from_bson(bson: BSON) do
+		for k, v in bson.to_json do self[k] = v
+	end
+
+	# Returns a new BSON object from `self`.
+	#
+	# ~~~
+	# var obj = new JsonObject
+	# obj["age"] = 10
+	# obj["name"] = "Rick"
+	# obj["ELS"] = new JsonArray
+	# var bson = obj.to_bson
+	# assert bson.to_s == """{ "age" : 10, "name" : "Rick", "ELS" : [  ] }"""
+	# ~~~
+	fun to_bson: BSON do return new BSON.from_json(self)
+end
+
+# An error returned by the mongoc client.
+#
+# Within the client, if a method returns `false` or `null` it's more likely that
+# an error occured during the execution.
+# You can see the last error occured with `Sys::last_mongoc_error`.
+#
+# Errors are formated in BSON.
+extern class BSONError `{ bson_error_t * `}
+
+	# Logical domain within a library that created the error.
+	fun domain: Int `{ return recv->domain; `}
+
+	# Domain specific error code.
+	fun code: Int `{ return recv->code; `}
+
+	# Human readable error message.
+	fun message: String import NativeString.to_s `{
+		return NativeString_to_s(recv->message);
+	`}
+
+	redef fun to_s do return "{message} (code: {code})"
+end
+
+redef class Sys
+	# Last error raised by `monogdb::MongoClient`.
+	var last_mongoc_error: nullable BSONError = null
+end
+
+# The MongoClient is used to connect to the mongo server and send queries.
+#
+# Usage:
+#
+# ~~~
+# var client = new MongoClient("mongodb://localhost:27017/")
+# client.close
+# ~~~
+extern class MongoClient `{ mongoc_client_t * `}
+
+	# Connect to `uri`.
+	new(uri: String) import String.to_cstring `{
+		mongoc_init();
+		return mongoc_client_new(String_to_cstring(uri));
+	`}
+
+	# Close the connexion between the client and MongoDb.
+	#
+	# FIXME: how to handle destructors?
+	fun close `{
+		mongoc_client_destroy(recv);
+		mongoc_cleanup();
+	`}
+
+	# Gets server data.
+	#
+	# Returns `null` if an error occured. See `Sys::last_mongoc_error`.
+	#
+	# ~~~
+	# var client = new MongoClient("mongodb://localhost:27017/")
+	# assert client.server_status["process"] == "mongod"
+	# client.close
+	# ~~~
+	fun server_status: nullable JsonObject
+		import set_mongoc_error, JsonObject.from_bson, JsonObject.as nullable `{
+		bson_error_t error;
+		bson_t reply;
+		if(!mongoc_client_get_server_status(recv, NULL, &reply, &error)){
+			MongoDb_set_mongoc_error(recv, &error);
+			return NULL;
```

Comment:
You have to use: `return null_JsobObject();` here. There is a conflict between Nit's null and C's NULL on extern classes... so NULL doesn't work.